### ISSUE TITLE
Thesis: self-sovereign identity (SSI)

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#5576]
Current status:
- Full-time for 11 November onwards (first courses)
- ToDo: formulate Cum Laude scientific idea
- Re-usage of Superapp infrastructure with IPv8
- cooperation with ministry
- tamper-proof hardware: https://www.bunniestudios.com/blog/?p=5921
- Ask [Stephan Wong faculty member](https://www.tudelft.nl/eemcs/the-faculty/departments/quantum-computer-engineering/computer-engineering/staff/stephan-wong/) for his Computer Engineering expertise? 
- Idea: Build and trail a secure identity wallet with Passport and Euro contents
- Idea: use [OpenTitan silicon root-of-trust](https://docs.opentitan.org)
- idea: import quantum randomness through the open [GPIO pins](https://www.instructables.com/id/A-Raspberry-Pi-based-Truly-Random-Number-Generator/), latest [science](https://arxiv.org/abs/2001.10625) (too immature tech)
- Too hardware focused (e.g. Firmware flashing and OS compile debugging?
- Also involve the TBM people for the legally valid signature on behalf of a legal entity?
- Deadline: 1 Dec 2020 have a Problem Description in 1 page
- Deadline graduation: current rules Tuesday 31 Aug 2021 graduation (this has not changed)
- **ToDo:** Sicco check of cybersecurity assignment

**Comment by synctext** [comment]
[Open crypto hardware wallet Trezor](https://wiki.trezor.io/Developers_guide:Contributing)

**Comment by qstokkink** [comment]
If you're into crypto: another fun, open and very difficult problem in the space of Self-Sovereign Identities is that of credential **revocation**. You can get a taste of the problem space in these papers: [paper1](https://link.springer.com/chapter/10.1007/978-3-642-15317-4_28), [paper2](https://dl.acm.org/doi/abs/10.1145/2994620.2994625). You'll have to make sure revocation data propagates decentrally and that the revocation lists/sets are anonymized.

**Comment by synctext** [comment]
Are we interleaving requirement analysis and prototyping?
Discussed "Member States"+EC co-author objective.

**Comment by InvictusRMC** [comment]
Sicco's assignment request:

- Short assignment description (1-2 A4), including:
  - Research area
  - Knowledge gap
  - Research questions

**Comment by synctext** [comment]
Today is the first thesis day!
Goal: determine direction in 3 weeks. Draft problem description.
Firmware project or algorithm work?
Related work on SSI: https://repository.tudelft.nl/islandora/object/uuid%3A200f1df0-adda-47a1-894c-baf54133035a
plus obviously prior thesis from our Lab: https://repository.tudelft.nl/islandora/object/uuid%3Aaab1f3ff-da54-47f7-8998-847cb78322c8

**Comment by InvictusRMC** [comment]
WIP proposal:

[Thesis_Proposal.pdf](https://github.com/Tribler/tribler/files/5588327/Thesis_Proposal.pdf)

**Comment by synctext** [comment]
Currently have a few possible directions:
- International Deployment, European-ID
- Identity Theft (private key compromise)
-  "industrial-strength SSI implementation". Native Kotlin of IPv8 zero-knowledge proof and from-scratch interaction model / GUI design. Trail with real users and high-end publication (IEEE Magazine level).

**Comment by qstokkink** [comment]
If you plan on investigating revocation, you may also want to take note of this recent MSc thesis (from the crypto group) regarding a credential revocation scheme: http://resolver.tudelft.nl/uuid:871e17aa-a023-461f-88b1-80d83c21de92

**Comment by InvictusRMC** [comment]
I have written the following Thesis proposal with the _Industry-Strength SSI_ in mind. I plan to submit this to Sicco.

[Thesis_Proposal_V2.pdf](https://github.com/Tribler/tribler/files/5592585/Thesis_Proposal_V2.pdf)

**Comment by synctext** [comment]
Nice! Seems ready for submission to Sicco. Minor comments:

> realised, there is still much to gain.

Still much left to investigate.

(4) interoperability is another key challenge. This needs to be an open standard and have an open reference implementation. But somehow the Spanish wallet provider needs to be able to talk fluently to Polish verifiable claim providers.
So, a German/Dutch international SSI trail would be a key milestone to achieve. 

> The main knowledge gap currently existing in the research area of SSI is the gap between the theoretical frameworks and the practicality of an implementation of these theoretical frameworks.

The main knowledge gap currently existing in the research area of SSI is the gap between the theoretical frameworks and the feasibility these theories. For instance, strict processing latency requirements on mobile devices. Governments would be reluctant to ask many of the European Union citizens to upgrade their phones. Governance requirements by nation-states, complexity, communication overhead, and fault-tolerance are other gaps.

> we can derive the following sub-questions that can be used to substantiate an answer to this research question

Too generic. We know prior work, functional requirements and its limits. This deviates from what is really needed. As scientists we can ignore the eIDAS and GDPR legal constraints somewhat. The next stage in the maturing of the SSI theoretical concepts is to go beyond mock-up trails. We need to devise _the most simple as possible_ SSI architecture which integrates the following required elements in a technology neutral manner:
- A wallet to store verifiable claims
- open API towards any commercial vendor of biometric authentication technology ([DONE](https://github.com/Tribler/py-ipv8/issues/74))
- open API towards any secure hardware token with challenge-response protocol to minimise side-channel leakage.
- Compatible with major deployed smartphone Operating Systems (Android,iPhone)
- Integrate a "trusted registry" of entities for which claims are accepted

**Comment by InvictusRMC** [comment]
Great feedback. I've updated the research questions and added a minor section on interoperability. The research questions could  still use a little refinement, however, are sufficient for submission in my opinion.

[Thesis_Proposal.pdf](https://github.com/Tribler/tribler/files/5597435/Thesis_Proposal.pdf)

**Comment by InvictusRMC** [comment]
Update:
* Got (almost) everything in order for BZK:
  - Will (most likely) be attending meeting next monday on their next plans for SSI.
  - Got access to their systems so I can freely communicate within BZK.
* Identified around 25 articles/whitepapers/theses to read:
  - Still have 19 to go through
* Finished Q1 courses (still had an oral & project to complete):
  - Still waiting on a single grade
  - However, Cum Laude average is secure

**Comment by synctext** [comment]
https://graz.pure.elsevier.com/en/publications/revocable-and-offline-verifiable-self-sovereign-identities

**Comment by InvictusRMC** [comment]
Had a great talk with @qstokkink. Got an introduction to the Python code and pointers to interesting authors and what literature to stay away from.

Also my literature list for now (will be growing after the pointers from Quinten): 

```
Stokkink, Q., Epema, D., & Pouwelse, J. (2020). A Truly Self-Sovereign Identity SystemarXiv preprint arXiv:2007.00415.

Dong, C., Wang, Z., Chen, S., & Xiang, Y. (2020). BBM: A Blockchain-Based Model for Open Banking via Self-sovereign Identity. In International Conference on Blockchain (pp. 61–75).

Stokkink, Q., & Pouwelse, J. (2018). Deployment of a blockchain-based self-sovereign identity. In 2018 IEEE International Conference on Internet of Things (iThings) and IEEE Green Computing and Communications (GreenCom) and IEEE Cyber, Physical and Social Computing (CPSCom) and IEEE Smart Data (SmartData) (pp. 1336–1342).

Zwitter, A., Gstrein, O., & Yap, E. (2020). Digital Identity and the Blockchain: Universal Identity Management and the Concept of the “Self-Sovereign” IndividualFrontiers in Blockchain, 3.

Van Bruggen, C.. (2020). Forward-looking consistency inAttribute-Based Credentials. 

Ferdous, M., Chowdhury, F., & Alassafi, M. (2019). In search of self-sovereign identity leveraging blockchain technologyIEEE Access, 7, 103059–103079.

Tobin, A., & Reed, D. (2016). The inevitable rise of self-sovereign identityThe Sovrin Foundation, 29(2016).

Cameron, K. (2005). The laws of identityMicrosoft Corp, 5, 8–11.

Allen, C.. (2016). The Path to Self-Sovereign Identity. 

Mühle, A., Grüner, A., Gayvoronskaya, T., & Meinel, C.. (2018). A survey on essential components of a self-sovereign identity. 

Zhou, T., Li, X., & Zhao, H. (2019). EverSSDI: Blockchain-based framework for verification, authorisation and recovery of self-sovereign identity using smart contractsInternational Journal of Computer Applications in Technology, 60(3), 281–295.

Coelho, P., Z\'uquete, A., & Gomes, H. (2018). Federation of Attribute Providers for User Self-Sovereign IdentityJournal of Information Systems Engineering & Management, 3(4).

Helmy, N.. (2020). Overview of Decentralized Identity Standards. 

Der, U., Jähnichen, S., & Sürmeli, J. (2017). Self-sovereign identity - opportunities and challenges for the digital revolutionarXiv preprint arXiv:1712.01767.

Wang, F., & De Filippi, P. (2020). Self-Sovereign Identity in a Globalized World: Credentials-Based Identity Systems as a Driver for Economic InclusionFrontiers in Blockchain, 2, 28.

Othman, A., & Callahan, J. (2018). The Horcrux Protocol: A Method for Decentralized Biometric-based Self-sovereign Identity. In Proceedings of the International Joint Conference on Neural Networks. Institute of Electrical and Electronics Engineers Inc..

Stokkink, Q., Epema, D., & Pouwelse, J. (2020). A Truly Self-Sovereign Identity SystemarXiv preprint arXiv:2007.00415.

Dong, C., Wang, Z., Chen, S., & Xiang, Y. (2020). BBM: A Blockchain-Based Model for Open Banking via Self-sovereign Identity. In International Conference on Blockchain (pp. 61–75).

Stokkink, Q., & Pouwelse, J. (2018). Deployment of a blockchain-based self-sovereign identity. In 2018 IEEE International Conference on Internet of Things (iThings) and IEEE Green Computing and Communications (GreenCom) and IEEE Cyber, Physical and Social Computing (CPSCom) and IEEE Smart Data (SmartData) (pp. 1336–1342).

Zwitter, A., Gstrein, O., & Yap, E. (2020). Digital Identity and the Blockchain: Universal Identity Management and the Concept of the “Self-Sovereign” IndividualFrontiers in Blockchain, 3.

Van Bruggen, C.. (2020). Forward-looking consistency inAttribute-Based Credentials. 

Ferdous, M., Chowdhury, F., & Alassafi, M. (2019). In search of self-sovereign identity leveraging blockchain technologyIEEE Access, 7, 103059–103079.

Tobin, A., & Reed, D. (2016). The inevitable rise of self-sovereign identityThe Sovrin Foundation, 29(2016).

Cameron, K. (2005). The laws of identityMicrosoft Corp, 5, 8–11.

Allen, C.. (2016). The Path to Self-Sovereign Identity. 

Mühle, A., Grüner, A., Gayvoronskaya, T., & Meinel, C.. (2018). A survey on essential components of a self-sovereign identity. 

Zhou, T., Li, X., & Zhao, H. (2019). EverSSDI: Blockchain-based framework for verification, authorisation and recovery of self-sovereign identity using smart contractsInternational Journal of Computer Applications in Technology, 60(3), 281–295.

Coelho, P., Z\'uquete, A., & Gomes, H. (2018). Federation of Attribute Providers for User Self-Sovereign IdentityJournal of Information Systems Engineering & Management, 3(4).

Helmy, N.. (2020). Overview of Decentralized Identity Standards. 

Format techreport not supported yet!

Der, U., Jähnichen, S., & Sürmeli, J. (2017). Self-sovereign identity - opportunities and challenges for the digital revolutionarXiv preprint arXiv:1712.01767.

Wang, F., & De Filippi, P. (2020). Self-Sovereign Identity in a Globalized World: Credentials-Based Identity Systems as a Driver for Economic InclusionFrontiers in Blockchain, 2, 28.

Othman, A., & Callahan, J. (2018). The Horcrux Protocol: A Method for Decentralized Biometric-based Self-sovereign Identity. In Proceedings of the International Joint Conference on Neural Networks. Institute of Electrical and Electronics Engineers Inc..

Speelman, T. (2020). Self-Sovereign Identity: Proving Power over Legal Entities. (Master Thesis, TU Delft).

Baars, D. (2016). Towards self-sovereign identity using blockchain technology. (Master Thesis, UTwente).
```

**Comment by InvictusRMC** [comment]
Some updates:

- Proposal cleared by both Sicco and Stjepan
- Cybsec has a requirement that I have to attend some presentations (will be taken care of by me)
- Currently still reading papers, hope to be done by end of week. Currently summarizing them for my own convenience and have used relevent literature to draft a properties/requirements and problem statement/background section.
- Have some nifty ideas we can discuss tomorrow (regarding a revocation method that can be performed relatively offline - atleast, requires less interactivity).
- Next steps:
  - Want to start implementing some code end of month.

[Here](https://github.com/Tribler/tribler/files/5704769/Thesis.pdf) is my very (!) rough draft. Thought I'd include it for completeness sake.

**Comment by synctext** [comment]
Lots of lost or stolen ID documents: https://nos.nl/nieuwsuur/artikel/2292728-we-verliezen-massaal-onze-paspoorten-fraudemeldingen-verdubbeld.html

**Comment by InvictusRMC** [comment]
I have made a design on how the _Hybrid-Revocation Model_ could work:
![hybrid-revocation](https://user-images.githubusercontent.com/22516119/103001729-d7b8f600-452d-11eb-8e32-49e40e18688d.png)

Also a design how it can be integrated in the regular verification flow:
![Claim-Verification-Flow](https://user-images.githubusercontent.com/22516119/103001864-1f3f8200-452e-11eb-930b-e756e644f869.png)

**Comment by InvictusRMC** [comment]
Update

- Designed diagrams for workflows
- Finished reading for now (related work section is still WIP)
- Started porting the current SSI code of py-ipv8 to Kotlin (nearing 1000 LoC)
  - Primary focus: 
    - Wallet Community (Self-imposed deadline: this month)
  - Secondary focus:
    - Identity Community (Self-imposed deadline: this month)
  - Plans after above tasks are completed:
    - Formalize own additions
    - Implement

**Comment by synctext** [comment]
- what are the field trail _key requirements_?
- what is the feature outcome of January sprint? 
- side-channel leakage, our Tor-like proxies are out of scope
- Current analysis: complexity of a 5-person field trail seem manageable.

**Comment by InvictusRMC** [comment]
Update:
* Continuing with engineering:
  * Attestation community almost done.
  * Wallet community should only be a few days work after that.
* Have everything running in an Android app in first week of February.
* Still need to have meeting with André about ESSIF requirements.
* Within BZK: will be having communication with Tim, Pim, and Quinten for international SSI tracks from EU.

**Comment by synctext** [comment]
- zero-knowledge proof Kotlin?
- @tim-w credentials for "Trustchain Superapp" currently on Android Play Store, obtain live crash reports.
  - Pull request and take over project?

**Comment by InvictusRMC** [comment]
### Progress Update
* AttestationCommunity functioning
  * Squished a ton of bugs (Tip to self: Kotlin docs are not always correct)
  * Did a huge number of tests:
    * Core crypto classes now have tests (which also pass)
    * Manual more e2e-like testing
  * Skipped the `irmaexact` attestation algorithm --> is also not fully implemented in Py-IPv8
  * Still have to implement `pengbaorange` (not much work, builds on Bonehexact which is 100% implemented)
* In order to prove that AttestationCommunity is functioning:
  * Built a concept GUI app on Android which shows the request of attestation signing and the storage of verifiable claims
  * Built a concept CLI app which shows the verify claim flow (this will also end up in the Android app, ofcourse)
  * Code base is now in Kotlin-IPv8 (as the Superapp is a bit bloated, this has faster development time)
    * Ensured that the code is functioning in the Superapp, so should be a matter of `ctrl+c ctrl+v` to get it up and running
      * Unfortunate side effect: dependency "hell" breaks Blockchain Engineering "Luxery Communism" project --> will fix this later on.
* Unfortunately, a lot of time went into seting up the Android app (tip to self: JDK8 source cannot be compiled with JDK11)
  * Ofcourse, this is a one time thing
* My next steps (this week):
  *  Finish IdentityCommunity
    * This community adds identity management
    * Complexity-wise, this code will take far less time
    * Goal: up and running this week
  *  Add `pengboarange` algorithm: tomorrow
* Next steps (sprint):
 * Formalise offline-first design requirements
 * Android engineering time:
   * Implement IdentityCommunity in Android
   * QR support
   * Keep everything quite simple, focus on functionality (in Android)
* EBSI/ESSIF link

**Comment by synctext** [comment]
- 2 March 2021: possible real trail with EuroTokens, Digital Basic Identity and 18+ attestations.
- Making solid progress, towards MvP
- international trail, live usage, and international co-authors becoming realistic

**Comment by InvictusRMC** [comment]
Progress update:

- Implemented Peng & Bao ZKP:
  - Verification and attestation fully working
- Fixed active verification for Boneh algorithm
- Implemented Trusted Authority Manager for IPv8
 - Stores the hash of the public key of authorities
 - Saved to database
- Extended attestation flow to incorporate signature:
  - Performed over metadata and serialized attestation
  - Allows for verification with Trusted Authorities
- Working on Android demo.

![image](https://user-images.githubusercontent.com/22516119/108997676-2d319080-76a0-11eb-9aa2-570650899379.png)
 
![image](https://user-images.githubusercontent.com/22516119/108997689-30c51780-76a0-11eb-91aa-e1a38adee1e7.png)

**Comment by InvictusRMC** [comment]
[Corona test app](https://tweakers.net/nieuws/178462/nederland-gaat-over-drie-weken-beginnen-met-app-die-negatieve-coronatest-toont.html) from the government:

**Comment by synctext** [comment]
- We have a screen and 18+ icon!  :
- Getting ready for 2 March or later trail with real digital :money_mouth_face: Euros and real age verification
- related work for broader perspective: scalability and private key storage are: ["Why blockchain is not yet working"](https://news.ycombinator.com/item?id=26130912)
- related work, Europe is forming real-time data portability if identity is solved, https://www.eff.org/deeplinks/2020/12/eus-digital-markets-act-there-lot-room-improvement

**Comment by synctext** [comment]
Related work by Evernym from Sovrin which uses central identity resolver architecture. Big 54MByte .APK on Android. "Connect.Me also includes an integration with the identity verification company Onfido." "[Using Connect.Me, individuals can:](https://www.evernym.com/blog/connect-me-sovrin-digital-wallet/)
- Form private, pairwise cryptographic connections with other entities in the Sovrin ecosystem
- Gather digital credentials
- Present digital proofs of part or all of your credentials, privately and securely
- Store (and restore from) a backup of your digital wallet wherever you want
- Answer secure messages from any connection you have"
All central IBM server infrastructure: "The core credential exchange technology that powers Connect.Me is being adopted on a global scale, via Hyperledger Indy, Hyperledger Ursa, and most recently, Hyperledger Aries." Look good! :gem: 
![image](https://user-images.githubusercontent.com/325224/109121163-33c71300-7747-11eb-891b-2dff642f71f4.png)

**Comment by InvictusRMC** [comment]
The basic functionality for the demo is done! 
[ssi-debug-apk-v0.1.zip](https://github.com/Tribler/tribler/files/6046252/ssi-debug-apk-v0.1.zip)
The app can be used as follows:

### The home screen
<Img width=350 src="https://user-images.githubusercontent.com/22516119/109230280-e5f0f000-77c4-11eb-9953-3c2cd3371358.png" />

The home screen shows a QR code of your public key. This can be scanned using the scanner to add this key to your trusted authorities. Secondly, the screen shows your attestations (currently only the Peng & Bao attestation). You can press on any of the attestations to display the QR code. This looks as follows:

<Img width=250 src="https://user-images.githubusercontent.com/22516119/109230538-539d1c00-77c5-11eb-8e26-86e6693a3456.png" />

This QR code can also be scanned using the scanner. 

### The scanner
The scanner has two functionalities: it can scan attestations and public keys. When scanning an attestation, the user is prompted the following:

<Img width=250 src="https://user-images.githubusercontent.com/22516119/109231355-bba03200-77c6-11eb-8c5f-d033eae0f19e.png">

If the user chooses to verify, the attestation is verified locally through the signature of the signee.

The results in one of the following results:

<Img width=200 src="https://user-images.githubusercontent.com/22516119/109231446-ea1e0d00-77c6-11eb-9b0c-c9e297b8f305.png" />
<Img width=200 src="https://user-images.githubusercontent.com/22516119/109231454-edb19400-77c6-11eb-9239-699db4790cd9.png" />

When scanning a public key, the user is prompted by the following and can add a new trusted authority:

<Img width=250 src="https://user-images.githubusercontent.com/22516119/109231555-1c2f6f00-77c7-11eb-97db-080129728169.png" />

### Users screen
On the users screen, attestations can be requested. When pressing on a valid peer, a prompt opens for an attestation name:

<Img width=250 src="https://user-images.githubusercontent.com/22516119/109231660-48e38680-77c7-11eb-9141-6653c493dca1.png" />

On the other client's end, the following prompt launched for entering a value (for now only numerical for the 18 plus check):

<Img width=250 src="https://user-images.githubusercontent.com/22516119/109231776-75979e00-77c7-11eb-8e8c-d523280fb723.png" />

A Peng Bao attestation is subsequently send to the requesting party and added to his wallet. This can then be presented and scanned as explained prior.

**Comment by InvictusRMC** [comment]
### Update:

- App redesign + features:
  - U.I. has now more actions through FABs
  - QR code generation is faster
  - Ability to delete attestations & authorities
  - Ability to request attestation through QR code
  - User is prompted for intent when scanning public key QR code:
    - Add as authority
    - Request attestation from peer with said public key
- `Kotlin-IPv8` merged in Master
- `Super-app` PR open
- Had contact with Wessel Blokzijl

**Comment by qstokkink** [comment]
Just had a quick glance at your source code: you may want to look into porting the [CommunicationManager](https://github.com/Tribler/py-ipv8/blob/master/ipv8/attestation/communication_manager.py) for a clean programmatic pseudonym management API, instead of accessing all of the primitives and database data directly. This class also includes rendezvous logic so you can find your counterparty on the Internet if more than 30 people join your community.

Side note: [this is how we map user interactions to the CommunicationManager in Python](https://github.com/Tribler/py-ipv8/blob/master/ipv8/REST/identity_endpoint.py).

**Comment by InvictusRMC** [comment]
> Just had a quick glance at your source code: you may want to look into porting the [CommunicationManager](https://github.com/Tribler/py-ipv8/blob/master/ipv8/attestation/communication_manager.py) for a clean programmatic pseudonym management API, instead of accessing all of the primitives and database data directly. This class also includes rendezvous logic so you can find your counterparty on the Internet if more than 30 people join your community.
> 
> Side note: [this is how we map user interactions to the CommunicationManager in Python](https://github.com/Tribler/py-ipv8/blob/master/ipv8/REST/identity_endpoint.py).

Thanks for the pointers, @qstokkink! The communication manager was definitely on my TODO list, however, I made the decision that, merely, the WalletCommunity is sufficient for the upcoming field trial of ~15 people.
> This class also includes rendezvous logic so you can find your counterparty on the Internet if more than 30 people join your community.

This I did not know and was something that already crossed my mind. This will definitely be necessary, I will port these. Thanks!

**Comment by InvictusRMC** [comment]
I just made a release for the 18 plus SSI app for the field trial this Thursday. [Download link](https://github.com/InvictusRMC/trustchain-superapp/releases/tag/demo-v0.1)

**Comment by InvictusRMC** [comment]
Nifty functionality: got attestation requesting and signing (and also the QR local verification) fully functional "offline" through Bluetooth! It's a bit slower, due to lower MTU, and unstable, but it does work.

**Comment by InvictusRMC** [comment]
### Progress update:
- Implemented `CommunicationManager` & `IdentityCommunity`: 
  - As mentioned by Quinten, includes rendezvous logic in order to locate peers when the IdentityCommunity grows too large.
    - Works by creating new channels, which basically function as sub-communities. Special rendezvous tokens allows other peers to discover these new channels when interacting with a client within such a channel.
- Implementing REST API in Kotlin: 
  - Similarly to the REST functionality in Python, this allows clients to interact with each other via the HTTP protocol.
  - Allows for an accessible interface as the functionality underneath becomes basically a black bocks obscured by the GET/POST requests. Should be finished this week.
  - RESP API simplifies ESSIF interoperability.
  - Example ESSIF:
    <details>
      <summary>Example ESSIF ID</summary>
      <p>
    
        ```json
        {
          "@context": [
            "https://www.w3.org/2018/credentials/v1",
            "https://essif.europa.eu/schemas/vc/2019/v1",
            "https://essif.europa.eu/schemas/eidas/2019/v1"],
          "id": "did:ebsi-eth:00000001/credentials/1872",
          "type": ["VerifiableCredential", "EssifVerifiableID"],
          "issuer": "did:ebsi-eth:00000001",
          "issuanceDate": "2019-06-22T14:11:44Z",
          "credentialSubject": {
            "id": "did:ebsi-eth:00000002",
            "currentFamilyName": "Franz",
            "currentGivenName": "Hinterberger",
            "dateOfBirth": "1999-03-22T00:00:00Z",
            "placeOfBirth": "Salzburg, Austria"
          },
          "proof": [ {
            "type": "EcdsaSecp256k1Signature2019",
            "created": "2019-06-22T14:11:44Z",
            "proofPurpose": "assertionMethod",
            "verificationMethod": "did:ebsi-eth:00000001#key-1",
            "jws": "eyJhbGciOiJSUzI1NiIsImI2NCI6ZmFsc2UsImNyaXQiOlsiYjY0Il19..TCYt5X
              sITJX1CxPCT8yAV-TvkIEq_PbChOMqsLfRoPsnsgw5WEuts01mq-pQy7UJiN5mgRxD-Wuc"
          } ]
        }
        ```
    
      </p>
    </details>

   - REST API talks in JSON format --> perfect for making SSI on IPv8 compatible with ESSIF on EBSI.
- Will be working on SSI blogpost for RvIG: general SSI information, albeit can be useful as practice for thesis.

**Comment by synctext** [comment]
Keeping track of developments
- SSI Toolbox for NL with IRMA integration (Pim Otte, Tim Speelman)
- [European SSI has open documentation](https://ec.europa.eu/cefdigital/wiki/pages/viewpage.action?pageId=262505750)
- International trail, industry-grade (secondary issues: loss recovery, revocation, theft locking, standardisation :3rd_place_medal:)

**Comment by qstokkink** [comment]
In case you care about IRMA: I recently fixed [the script that pulls BRP attributes from the Nijmegen IRMA server](https://github.com/Tribler/py-ipv8/blob/master/ipv8/attestation/wallet/irmaexact/enroll_script.py). I'm not sure if I'd recommend supporting this, as you need to go through a DigiD portal.. but it works again.

**Comment by InvictusRMC** [comment]
Progress update:
 
- Implemented greedy revocation:
  - New `GossipCommunity`
  - Peers send each other periodically `PreviewPayloads`:
    - Contain key-value pairs of known public key hashes and latest revocation versions
    - Received peers can send back an `UpdateRequestPayload`:
  - `UpdateRequestPayloads`: Contain key-value pairs of requested updates and the lowest version the requesting peer knows.
    - Receiving peer sends back `UpdateChunkPayloads`:
  - `UpdateChunkPayloads`:
    - The revocations belonging to each key-value pair of public key hash and version are fragmented into a stream of UDP packets and reconstructed at the requesting clients end
    - The stream of UDP packets thus belong to a single version and public key hash --> minimizes impact in case of packet loss (firstly implemented by sending all updates in a single stream --> 100% loss in case 1 packet was lost)
  - The availability issue:
    - Most favorably, we want revocations to be known all across the network
    - This led to the design choice of making all peers propagate all revoked signatures
    - Peers can only verify signatures of the revocations in case the public key is known to them:
      - Public key is not sent along (aids privacy + reduced network usage)
    - Peers thus insert revocations without checking in case the public key belonging to the hash is not known to them, else they firstly validate the signature.
    - Possible issue: spammers can insert random revocations into clients databases, solutions:
      - Clients block spammers (make the issue resolve itself)
      - Send along public keys to verify revocations (does not really solve the issue as we do not necessarily know to whom the public key belongs, just makes it a bit harder as a spammer requires multiple keys)
      - Do not insert revocations of unknown authorities (leads to lower availability, although, greatly reduces storage overhead and network usage).
  - The "greediness":
    - Clients request all versions they do not know from possibly multiple peers at ones:
      - Updates it already learned in the time between requesting and receiving are ignored
  - The rough edges of this design:
    - Concurrency constraint: updates are ignored in case the previous version is not known: fixed in `eventually consistent` fashion: i.e., the next `PreviewPayload` will give the client another chance at gaining the packages in order.
    - In case a client requests version x, however, the other party learned of versions `[x+1, x+y]` meanwhile, the other party also sends these along as it has no true notion of the requested versions (requesting party only sends there lowest known version along). The receiving party now drops any of the versions newer than `x` as it did not request these --> counter measure against spammers.

**Comment by synctext** [comment]
- Stable code for 8 April presentation on App Store
- Bulk data sharing using Libtorrent works in MusicDAO

**Comment by InvictusRMC** [comment]
Another paper by Sovrin on revocation: https://sovrin.org/wp-content/uploads/AnonCred-RWC.pdf

**Comment by InvictusRMC** [comment]
Just had an idea for the secondary issue of loss recovery:

* Main issue: private key recovery --> Use Shamir's Secret Sharing with trusted persons (e.g. x close friends or even your local government). Next, we can use this private key recovery mechanism in two ways:
1. First method:
   - (1) Periodically encrypt serialized wallet contents with private key as encryption key and (2) store on an external storage solution (e.g. some cloud storage provider).
   - Use recovered private key to decrypt and restore encrypted wallet backup.
   - (Wallet is stored encrypted as this gives some guarantees on confidentiality and integrity).
2. Second method:
   - (1) Use recovered private key in challenge mechanism to prove ownership over attestations and (2) make each authority resend the attestations (trivial solution).

**Comment by qstokkink** [comment]
You could check out the results of #3245 and #3246. Both groups used Shamir shares for recovery.

**Comment by synctext** [comment]
WEF authors to contact: https://www.weforum.org/agenda/2021/04/trust-is-fraying-how-trust-anchors-could-help-gtgs/

**Comment by InvictusRMC** [comment]
### Update Progress
* Code base:
  - PRs open for 18 days
  - Extensive (manual) e2e testing for `CommunicationManager`:
    - Implements rendezvous logic:
       -  Allows peers to create a subcommunity (second instance of a Community) on runtime 
       - --> No more visibility issues
       - Works via a special code a client can generate themselves and distribute via other means.
* Revocation:
  - Not developed further since last meeting, but
  - [Khovratovich & Jason (2017)](https://sovrin.org/wp-content/uploads/AnonCred-RWC.pdf) state three methodologies for revocation specifically for SSI:
    1. Signature Lists [(Nakanishi et al., 2009)](https://www.iacr.org/archive/pkc2009/54430468/54430468.pdf). Paper on revoking keys, where keys of non-revoked members are fetched.
    2. Cryptographic accumulators [(Camenisch et al., 2009)](https://eprint.iacr.org/2008/539.pdf). Hashes all valid credentials to a single (accumulated) value. A client uses a witness to prove that the accumulator comprises their identity. Solution requires interactivity with issuer and witnesses.
    3. Forward revocation lists [(Verheul, 2016)](https://img.chainnews.com/paper/e56be00157a19f11a3a8007a8b427103.pdf). Issuer generates revocation entries, client present non-revocation proof. A verifier tests this proof against every entry.
* Possible novelty in revocation:
  - Distributed: most (all?) of the solutions do not use other clients as relay stations for revoked signatures. Gossip can be a novelty.
  - Usage of bloom filters for speed optimization. My current solution uses a revocation list of the hashes of revoked signatures. Bloom filters are extremely efficient datastructures which help us say with certainty that a signature is not revoked. In case the bloom filter reports that there might be a match, then we can perform the `O(n)` search. This `O(n)` search can also be optimized by using a smart search algorithm (e.g Heapsort worst case `O(n log n)`.
  - There are some papers on bloom filters and revocation. (Did not yet dive in):
    - (https://ieeexplore.ieee.org/abstract/document/7192615)
    - (https://ieeexplore.ieee.org/abstract/document/5719271)
    - (https://ieeexplore.ieee.org/abstract/document/8288287) 

* Article:
  - Started working on article. ~(4-5 pages now, possibly too in-depth, but fragments can be used for the thesis itself)
  - What results to discuss in article?
  - [Article_Draft_19-04-21.pdf](https://github.com/Tribler/tribler/files/6333967/Article_Draft_19-04-21.pdf)

**Comment by synctext** [comment]
- Science in thesis?  ?Distributed and practical revocation systems for passport-grade identity
- Related work - credentials and web-of-trust: [Trust over IP foundation](https://trustoverip.org/wp-content/uploads/sites/98/2020/05/Introducing-The-Trust-over-IP-Foundation-V1.pdf)

**Comment by InvictusRMC** [comment]
Progress update:

* App (new logic implemented):
  - [x] Ability to set special rendezvous token
  - [x] Ability to request attestation whilst locating the other peer through rendezvous token
  - [x] Requests screen in order answer to attestation and verification requests
  - [x] Ability to sign and verify small (!) images --> can be used for ID pictures, idea to write something about this in thesis/article
  - [ ] UI component to see verification output
  - [ ] UI component to revoke signatures
* Writing:
  - Started writing on design of revocation, becoming more clear that fully distributed (offline usable) revocation has not been done yet
* SSI design:
  - Working on attestation presentation logic for "Von Der Leyen" trail
  - Significantly improved QR code presentation (for offline attestation verification):
     - Less vulnerable to replay attack 
     - Ability to split large attestation into two QR codes
     - QR generation way quicker
* BZK:
  - Next week presentation on SSI & my thesis

Proposed roadmap (next two weeks):
- [ ] Be done with app this week (including "Von Der Leyen" presentation logic)
- [ ] Revocation mechanism "beyond state of art" analysis/comparison
  - Have this written down in both article and thesis

**Comment by synctext** [comment]
- more scientific terms, not secondary something. ```Secure communication primitive```
- How different is your unique code from standard TLS?

**Comment by InvictusRMC** [comment]
"Von Der Leyen"-Trial (Although name is revealed, 18 plus also works)

https://user-images.githubusercontent.com/22516119/117708448-1a732600-b1d0-11eb-8d66-841b3da1bb81.mp4

**Comment by InvictusRMC** [comment]
### Progress

- Finished app:
  - Revocation logic implemented and functional
  - Presentation logic & answers to request for specific attributes ("Von Der Leyen"-Trial, see video in previous comment)
- Revocation functional:
  - Did analysis on convergence time (written down in Thesis)
  - Did analysis on impact of Bloom filters (written down in Article, should probably be moved to Thesis as a whole)
  - Working on runtime analysis graphs (bottleneck: only max ~10 clients before CPU is capped)
- Writing:
  - Have written more in-depth about design:
    - Explained protocol in terms of attestation signing and verification
    - Explained revocation
    - Extensive convergence time analysis
- Next steps:
  - Finish article this week/begin next week
  - Continue writing thesis
  - Investigate usage of accumulators versus hashes for Revocation
- General remarks:
  - Presentation on my Thesis @BZK on the 7th
  - Two PRs open
  - More happy with progress, but still require a lot of work on writing.
  - Need to plan in Thesis defence.

[DRAFT_Thesis_27-05-2021.pdf](https://github.com/Tribler/tribler/files/6554107/DRAFT_Thesis_27-05-2021.pdf)
[DRAFT_Article_27-05-2021.pdf](https://github.com/Tribler/tribler/files/6554108/DRAFT_Article_27-05-2021.pdf)

**Comment by synctext** [comment]
- How does your identity framework differ from FIDO2, OpenID, SAML?

**Comment by InvictusRMC** [comment]
With thanks to Andre:

https://eur-lex.europa.eu/legal-content/EN/TXT/?uri=COM%3A2021%3A281%3AFIN&qid=1622704576563

**Comment by synctext** [comment]
Please study this master thesis format for Cum Laude: https://github.com/Tribler/tribler/issues/5221#issuecomment-855911194

**Comment by InvictusRMC** [comment]
### Progress:

Article:
- Created table comparing other revocation algorithms
- Improved structure of article
- TODO:
  - Remove some less relevant information
  - Add section on scalability (still in progress of creating graphs)
Thesis:
  - Finished piece on Identity and SSI:
    - Philosophical nature of identity
    - Digital identity (introduced a small model describing the current ecosystem)
    - Evolution of digital identity
    - Problems in digital identity
    - Recorded the history of SSI for the first time (?) in academia
    - Discussed shortcoming of the term and proposed a new one
    - Discussed what SSI is through properties and proposed a new framework (needs perhaps another revision) aiding a definition of SSI.
  - TODO:
    - Write design piece on SSI (can be copied from article and then extended upon)
    - Write about ethics
    - Write about the lifecycle
    - Write about benefits for users
    - Write evaluation
 Test plan:
    - Need another day or two to finish test plan. Have drafted scenario's, but still need to strengthen them and define metrics.

[DRAFT_Article_13-06-2021.pdf](https://github.com/Tribler/tribler/files/6646023/DRAFT_Article_13-06-2021.pdf)
[DRAFT_Thesis_13-06-2021.pdf](https://github.com/Tribler/tribler/files/6646022/DRAFT_Thesis_13-06-2021.pdf)
[DRAFT_Test_Plan_14-06-2021.pdf](https://github.com/Tribler/tribler/files/6646086/DRAFT_Test_Plan_14-06-2021.pdf)

**Comment by InvictusRMC** [comment]
Thesis structure guidelines:
https://www.tudelft.nl/studenten/ewi-studentenportal/onderwijs/afstuderen-msc/thesis-report

**Comment by synctext** [comment]
Great article first draft! Please put the table first, as hook for your storyline that after 48 years since invention of public key crypto its not ready yet. Mark red (smaller) and green checks (bigger) plus clearly: _this work_.
btw Gemeente Hoorn crypto deployment for citizens: https://github.com/discipl/waardepapieren/projects/2

**Comment by InvictusRMC** [comment]
### Progress Update

* Spent mostly writing Thesis:
  - Have come a long way with main body. Document nearly 80 pages
* Article fixed table, still requires some removal of irrelevant/too detailed information.
  - Need to add results showcasing propagation time.
* Testing plan is now a single scenario showcasing both credentials and revocation.
* Have presentation about thesis this afternoon for Cybsec
* Had a talk with Prof Epema

[Thesis-Draft-30-60-21.pdf](https://github.com/Tribler/tribler/files/6738628/Thesis-Draft-30-60-21.pdf)
[Article-Draft-30-60-21.pdf](https://github.com/Tribler/tribler/files/6738629/Article-Draft-30-60-21.pdf)
[Testing-Plan-30-60-21.pdf](https://github.com/Tribler/tribler/files/6738630/Testing-Plan-30-60-21.pdf)

**Comment by synctext** [comment]
```
This chapter describes the overall outline of the thesis. Firstly, the research area is introduced, providing brief insights into the
topic of Self-Sovereign Identity. Next, the motivation for this research is discussed, touching upon current shortcomings and
 opportunities for SSI. Then, the research questions are discussed. Finally, the contributions of this work
are highlighted and the outline of the remainder of the document is given.
```
My preference is less boilerplate in a thesis and more condense formatting:
_We introduce the topic of Self-Sovereign Identity, discuss the motivation for this research, and outline the research questions of this thesis._

**Comment by InvictusRMC** [comment]
Progress Update:

- Focused on results and article in general
- Restructure in thesis

[Thesis-Draft-13-07-21.pdf](https://github.com/Tribler/tribler/files/6806948/Thesis-Draft-13-07-21.pdf)

**Comment by InvictusRMC** [comment]
[Thesis-Draft-23-07-21.pdf](https://github.com/Tribler/tribler/files/6867482/Thesis-Draft-23-07-21.pdf)

**Comment by synctext** [comment]
Simplified storyline:
- no "Industry-Grade", simply the first solution believe to be usable for mass usage because meets both primary and secondary requirements.
- no "Hybrid Revocation Model", simply the first fully distributed

**Comment by InvictusRMC** [comment]
[Article30.pdf](https://github.com/Tribler/tribler/files/6905588/Article30.pdf)

**Comment by synctext** [comment]
Offline usage is important: [flooding of cities results in Alipay going offline](https://www.scmp.com/news/people-culture/environment/article/3142544/china-floods-digital-dark-age-after-disaster-wreaks). Digital identities should be disaster-proof.

Or the de-platforming storyline, revocations are a collaborative alternative to Big Tech dictatorships of identity layer. Deplatforming: Your identity gets canceled by a profit-driven entity.
https://journals.sagepub.com/doi/full/10.1177/0267323120922066
```
Extreme, anti-establishment actors are being characterized increasingly as ‘dangerous individuals’ by the social media
platforms that once aided in making them into ‘Internet celebrities’. These individuals (and sometimes groups) are
being ‘deplatformed’ by the leading social media companies such as Facebook, Instagram, Twitter and YouTube for
such offences as ‘organised hate’. Deplatforming has prompted debate about ‘liberal big tech’ silencing free speech and
taking on the role of editors, but also about the questions of whether it is effective and for whom.
```

**Comment by InvictusRMC** [comment]
[Rowdy_Chotkan_MSc_Thesis_IG-SSI.pdf](https://github.com/Tribler/tribler/files/6998825/Rowdy_Chotkan_MSc_Thesis_IG-SSI.pdf)

As also submitted via mail.

**Comment by InvictusRMC** [comment]
[Slides WIP.pptx](https://github.com/Tribler/tribler/files/7045063/Defence-Slides.pptx)

**Comment by synctext** [comment]
[Msc Thesis DONE](https://repository.tudelft.nl/islandora/object/uuid:32711378-2f6f-452e-b65c-1866c471e934/datastream/OBJ/download)
